sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/OptimizeSkewedPartitions.scala

```diff
@@ -0,0 +1,297 @@
+    ThreadUtils.awaitResult(metrics, Duration.Zero)
+  }
+
+  /**
+   * Base optimization support check: the join type is supported and plan statistics is available.
+   * Note that for some join types(like left outer), whether a certain partition can be optimized
+   * also depends on the filed isSkewAndSupportsSplit.
+   */
+  private def supportOptimization(
+    joinType: JoinType,
+    leftStage: QueryStageExec,
+    rightStage: QueryStageExec): Boolean = {
+    val joinTypeSupported = supportedJoinTypes.contains(joinType)
+    val shuffleStageCheck = ShuffleQueryStageExec.isShuffleQueryStageExec(leftStage) &&
+      ShuffleQueryStageExec.isShuffleQueryStageExec(rightStage)
+    joinTypeSupported && shuffleStageCheck
+  }
+
+  private def supportSplitOnLeftPartition(joinType: JoinType) = {
+    joinType == Inner || joinType == Cross || joinType == LeftSemi ||
+      joinType == LeftAnti || joinType == LeftOuter
+  }
+
+  private def supportSplitOnRightPartition(joinType: JoinType) = {
+    joinType == Inner || joinType == Cross || joinType == RightOuter
+  }
+
+  private def estimatePartitionStartEndIndices(
+      mapOutputStatistics: MapOutputStatistics,
+      omittedPartitions: mutable.HashSet[Int] = mutable.HashSet.empty): Array[(Int, Int)] = {
+    val length = mapOutputStatistics.bytesByPartitionId.length
+    val partitionStartIndices = ArrayBuffer[Int]()
+    val partitionEndIndices = ArrayBuffer[Int]()
+    (0 until length).map { i =>
+      if (!omittedPartitions.contains(i)) {
+        partitionStartIndices += i
+        partitionEndIndices += i + 1
+      }
+    }
+    partitionStartIndices.zip(partitionEndIndices).toArray
+  }
+
+  def handleSkewJoin(plan: SparkPlan): SparkPlan = plan.transformUp {
+    case smj @ SortMergeJoinExec(leftKeys, rightKeys, joinType, condition,
+    SortExec(_, _, left: QueryStageExec, _),
+    SortExec(_, _, right: QueryStageExec, _))
+      if supportOptimization(joinType, left, right) =>
+      val leftStats = getStatistics(left)
+      val rightStats = getStatistics(right)
+      val numPartitions = leftStats.bytesByPartitionId.length
+
+      val leftMedSize = medianSize(leftStats)
+      val rightMedSize = medianSize(rightStats)
+      logDebug(s"HandlingSkewedJoin left medSize: ($leftMedSize)" +
+        s" right medSize ($rightMedSize)")
+      logDebug(s"left bytes Max : ${leftStats.bytesByPartitionId.max}")
+      logDebug(s"right bytes Max : ${rightStats.bytesByPartitionId.max}")
+
+      val skewedPartitions = mutable.HashSet[Int]()
+      val subJoins = mutable.ArrayBuffer[SparkPlan]()
+      for (partitionId <- 0 until numPartitions) {
+        val isLeftSkew = isSkewed(leftStats, partitionId, leftMedSize)
+        val isRightSkew = isSkewed(rightStats, partitionId, rightMedSize)
+        val isSkewAndSupportsSplit =
+          (isLeftSkew && supportSplitOnLeftPartition(joinType)) ||
+            (isRightSkew && supportSplitOnRightPartition(joinType))
```

Comment:
if both sides are skewed, we should be stricter about the join type, right?

Reply:
yes. If two sides are skewed, we need to further check the join type.

Reply:
where is the check?